src/trace/trace.cpp

```diff
@@ -489,28 +356,30 @@ otf2::writer::local& Trace::cpu_sample_writer(int cpuid)
 #else
     auto name = (boost::format("sample cpu %d") % cpuid).str();
 
-    // As the cpuid is unique in this context, create only one writer/location per tid
-    auto location = cpu_sample_locations_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(cpuid),
-        std::forward_as_tuple(location_ref(), intern(name), location_groups_cpu_.at(cpuid),
-                              otf2::definition::location::location_type::cpu_thread));
+    // As the cpuid is unique in this context, create only one writer/location per cpuid
+    const auto& location = registry_.emplace<otf2::definition::location>(
+        ByCpuSampleWriter(cpuid), intern(name),
+        registry_.get<otf2::definition::location_group>(ByCpu(cpuid)),
+        otf2::definition::location::location_type::cpu_thread);
 
-    return archive()(location.first->second);
+    comm_locations_group_.add_member(location);
+    return archive_(location);
 #endif
 }
 
 otf2::writer::local& Trace::cpu_switch_writer(int cpuid)
 {
-    // As CPU IDs are unique in this context, create only one writer/location per
-    // CPU ID
+    // As the cpuid is unique in this context, create only one writer/location per
+    // cpuid
     auto name = intern((boost::format("cpu %d") % cpuid).str());
 
-    auto location = cpu_switch_locations_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(cpuid),
-        std::forward_as_tuple(location_ref(), name, location_groups_cpu_.at(cpuid),
-                              otf2::definition::location::location_type::cpu_thread));
+    const auto& location = registry_.emplace<otf2::definition::location>(
+        ByCpuSwitchWriter(cpuid), name,
+        registry_.get<otf2::definition::location_group>(ByCpu(cpuid)),
+        otf2::definition::location::location_type::cpu_thread);
 
-    return archive()(location.first->second);
+    comm_locations_group_.add_member(location);
```

Comment:
Where does this line come from? 

Also, add_member is only an [emplace_back](https://github.com/tud-zih-energy/otf2xx/blob/master/include/otf2xx/definition/detail/group_impl.hpp#L122)